### PR TITLE
Add support for escaped chars in alpn parameters per per draft-ietf-dnsop-svcb-https-08

### PR DESCRIPTION
**Pull request by rafaelvanoni**

### CODE REVIEW

File: wdns/str_to_rdata_ubuf.c [hunk 1]
```diff
@@ -197,31 +197,65 @@ str_to_svcparam(ubuf *u, char *keyval)
...
+				c = src[i];
+				ubuf_append(v, &c, 1);
+				bytes++;
+			} else if (c == ',') {
+				ubuf_append(u, (uint8_t *)&bytes,
+				    sizeof (uint8_t));
+				ubuf_append(u, (uint8_t *)ubuf_data(v), bytes);
+
+				ubuf_reset(v);
+				val_len += bytes + sizeof(uint8_t);;
```
Comment (djw1149) [inline wdns/str_to_rdata_ubuf.c:241]: remove doubled semicolon

File: wdns/str_to_rdata_ubuf.c [hunk 2]
```diff
@@ -197,31 +197,65 @@ str_to_svcparam(ubuf *u, char *keyval)
...
+			if (c == '\\') {
+				/* shouldn't happen, but just to be safe.. */
+				if (++i >= len) {
+					return (wdns_res_parse_error);
+				}
+				c = src[i];
+				ubuf_append(v, &c, 1);
+				bytes++;
+			} else if (c == ',') {
+				ubuf_append(u, (uint8_t *)&bytes,
```
Comment (djw1149) [inline wdns/str_to_rdata_ubuf.c:236]: What happens if a comma is the first character we see? We'll append 0 to u, etc...
Maybe wrap a `if (bytes > 0` around lines 237-241?

**Comment by djw1149** [comment]
I don't entirely understand the I-D format requirements, so my review is not complete.

File: wdns/rdata_to_ubuf.c [hunk 3]
```diff
@@ -124,15 +124,32 @@ svcparam_to_str(uint16_t key, const uint8_t *src, uint16_t len, ubuf *u)
...
+			while (l--) {
+				uint8_t c = *ptr++;
+
+				if (c == '"') {
+				        ubuf_add_cstr(u, "\\\"");
+				} else if (c == '\\') {
+				        ubuf_add_cstr(u, "\\\\");
+				} else if (c == ',') {
+					ubuf_add_fmt(u, "\\,");
+				} else if (c >= ' ' && c <= '~') {
```
Comment (cmikk) [inline wdns/rdata_to_ubuf.c:147]: If `c == ' '` here, the output will trip up the `strtok_r(NULL, " \t\r\n", &endp)` parsing of the value. It would be more robust here to:
1. Always output a double-quoted (and escaped where necessary) string for `alpn` and default `keyNNN` values. 
2. Parse `alpn` and `keyNNN` values as possibly quoted character-string values (e.g., with `rdata_from_str_string()` into a separate ubuf, relying on `rdata_from_str_string()` to determine the end of the value)
3. For alpn values, parse the comma separated list of alpn values  from the result of (2)

The ipv[46]hint, port, and ech values can be (and are) encoded with a limited character set not requiring quoting, regardless of their wire format contents, which is why the standard can (and does) prohibit escape characters in them. I read the standard as allowing these to be quoted, but handling that is not strictly necessary. I expect wdns to be able to parse its own output back to the original rdata, and for that output to be parseable by other DNS zone file implementations, but do not expect wdns to parse all valid zone file contents.

File: wdns/str_to_rdata_ubuf.c [hunk 4]
```diff
@@ -197,31 +197,65 @@ str_to_svcparam(ubuf *u, char *keyval)
...
 		 * SvcParamValue. These pairs MUST exactly fill the
 		 * SvcParamValue; otherwise, the SvcParamValue is malformed.
 		 */
-		tok = strtok_r(NULL, ",", &endp);
+		char *src;
+		int len;
+		ubuf *v;
+		uint8_t c, bytes;
+
+		tok = strtok_r(NULL, " \t\r\n", &endp);
```
Comment (cmikk) [inline wdns/str_to_rdata_ubuf.c:213]: See comment on wdns/rdata_to_ubuf.c

This also affects the `default:` case of this switch statement, handling `keyNNN` values

**Review by cmikk — CHANGES_REQUESTED** [review]

File: wdns/rdata_to_ubuf.c (hunk 3 above)
Comment (rafaelvanoni) [reply wdns/rdata_to_ubuf.c:149]: Ok.. I think I addressed your comments. I'm wrapping alpn and keyNNN values around double quotes (in presentation format only). Please take a look.

Comment (cmikk) [reply wdns/rdata_to_ubuf.c:149]: The presentation format looks right with these latest changes, but the parsing needs work.  In particular, it will fail if any service parameter value contains a space.  The following patch to the rdata_to_str (+ round trip) test cases illustrates the remaining. This will require moving away from `strtok_r` on whitespace to separate service parameters, as that cannot account for quoted spaces, which will involve a larger refactor of the parsing.

`
--- a/t/test-rdata_to_str.c
+++ b/t/test-rdata_to_str.c
@@ -410,13 +410,13 @@ struct test tdata[] = {
                    "\x00\x01"          /* alpn in network order */
                    "\x00\x03"          /* length of the alpnid in net order */
                    "\x02\x68\x32"      /* length-value */
-                   "\x00\x09"          /* key9 in network order */
-                   "\x00\x03"          /* length of key9 in net order */
-                   "\x61\x6e\x79",     /* length-value */
-               .input_len = 17,
+                   "\x00\x08"          /* key8 in network order */
+                   "\x00\x09"          /* length of key9 in net order */
+                   "has space",        /* length-value */
+               .input_len = 23,
                .rrclass = WDNS_CLASS_IN,
                .rrtype = WDNS_TYPE_HTTPS,
-               .expected = "1 . alpn=\"h2\" key9=\"any\"",
+               .expected = "1 . alpn=\"h2\" key8=\"has space\"",
        },
 
        { 0 }
`

Comment (rafaelvanoni) [reply wdns/rdata_to_ubuf.c]: Ok, I re-wrote the parsing of alpn and arbitrary key values to account for escaped commas and blank spaces. Please take a look.

File: t/test-rdata_to_str.c [hunk 5]
```diff
@@ -334,18 +334,6 @@ struct test tdata[] = {
 		.expected = "1 foo.example.com. key667=\"hello\"",
 	},
 
-	{ /* appendix D, figure 5 */
```
Comment (djw1149) [inline t/test-rdata_to_str.c]: maybe leave in the code under an #ifdef pair with a comment why it does not work now?

File: wdns/rdata_to_ubuf.c [hunk 6]
```diff
@@ -245,11 +204,75 @@ svcparam_to_str(uint16_t key, const uint8_t *src, uint16_t len, ubuf *u)
...
+	 * The wire format value for "alpn" consists of at least one * "alpn-id"
+	 * prefixed by its length as a single octet, and these length-value
+	 * pairs are concatenated to form the SvcParamValue. These MUST exactly
+	 * fill the SvcParamValue; otherwise, the SvcParamValue is malformed.
+	 *
+	 * Note that we always wrap such values around double quotes in
+	 * presentation format.
+	 */
+	case spr_alpn:
+		ubuf_add_fmt(u, "\"");
```
Comment (djw1149) [inline wdns/rdata_to_ubuf.c:217]: I think if you are not using printf-style formatting, these ubuf_add_fmt() calls should be ubuf_add_cstr() calls. Which is only mildly important if we are doing a _lot_ of conversions, so maybe not worth changing the code.

File: wdns/str_to_rdata_ubuf.c [hunk 7]
```diff
@@ -131,28 +131,19 @@ str_to_ubuf(const char *str, ubuf *u, const char **retp)
 
 /*
  * str_to_svcparam() translates a "key=val" string to the SVCB/HTTPS wire
- * format as specified in draft-ietf-dnsop-svcb-https-08, section 2.2. Note
- * that some keys don't require a value while others can have more than one.
```
Comment (djw1149) [inline wdns/str_to_rdata_ubuf.c]: why did you remove the comment?  there are a few places here you reviewed what I thought we useful comments.

**Review by djw1149 — COMMENTED** [review]
just minor comments, can be ignored.

File: wdns/str_to_rdata_ubuf.c [hunk 8]
```diff
@@ -263,70 +233,137 @@ str_to_svcparam(ubuf *u, char *keyval)
...
+	 */
+	case spr_alpn:
+	default: {
+		uint8_t c, bytes;
+		bool quoted;
+		char *src;
+		int len, i;
+		ubuf *v;
+
+		src = strdup(val);
```
Comment (cmikk) [inline wdns/str_to_rdata_ubuf.c:280]: Is this `strdup()` necessary? I do not see either `src` or `val` being modified, so `val` could be `const char *` and used directly.

(Additionally, `src` is leaked in most `return (wdns_parse_error);` cases. If the `strdup()` is necessary, these leaks need to be fixed.)

File: wdns/str_to_rdata_ubuf.c [hunk 9]
```diff
@@ -934,22 +971,121 @@ _wdns_str_to_rdata_ubuf(ubuf *u, const char *str,
 		}
 
 		case rdf_svcparams: {
-			/* process ony key=val pair at a time */
-			char *buf, *tok, *endp = NULL;
+			char *start, *eol;
+			int prev_key = -1;
+
+			eol = (char *)str;
```
Comment (cmikk) [inline wdns/str_to_rdata_ubuf.c:977]: Casting `const char *` to `char *` should be avoided, as it bypasses an important safety measure in C.

That said, I do not see `start` or `eol` (or `key`, `val`, or `end` below) being used to modify the underlying string, so these should all be declared `const char *`.

File: wdns/str_to_rdata_ubuf.c [hunk 10]
```diff
@@ -934,22 +971,121 @@ _wdns_str_to_rdata_ubuf(ubuf *u, const char *str,
...
+				dup = strndup(val, end - val);
+
+				res = str_to_svcparam(u, k, dup);
 				if (res != wdns_res_success) {
-					goto err;
+					return (res);
 				}
 
-				tok = strtok_r(NULL, " ", &endp);
+				free(dup);
```
Comment (cmikk) [inline wdns/str_to_rdata_ubuf.c:1085]: This `free()` needs to be moved earlier to avoid leaking if `res != wdns_res_success`.

Note that if `str_to_svcparam` could avoid using `strtok_r` (similar parsing code elsewhere in wdns uses `strpbrk()` to non-destructively scan for separators), it could accept a `const char *` and make `strdup()`/`free()` pairs unnecessary.

File: wdns/rdata_to_ubuf.c [hunk 11]
```diff
@@ -503,6 +547,15 @@ _wdns_rdata_to_ubuf(ubuf *u, const uint8_t *rdata, uint16_t rdlen,
 					goto err;
 				}
 
+				if (key == spr_invalid) {
+					goto err;
+				}
+
+				/* param keys must be in ascending order */
```
Comment (cmikk) [inline wdns/rdata_to_ubuf.c:554]: While the draft spec requires keys to be sorted, for passive DNS purposes we should not reject these -- we can present the parameters in the order they appear on the wire (and conversely parse that presentation back to the original wire layout), and should.

**Review by cmikk — CHANGES_REQUESTED** [review]

File: wdns/str_to_rdata_ubuf.c [hunk 12]
```diff
@@ -263,70 +232,137 @@ str_to_svcparam(ubuf *u, char *keyval)
...
-		 * The rfc doesn't specify whether these types of keys can have
-		 * multiple values or not, so for now we treat it as a single
-		 * string (that could contain multiple comma separated values).
-		*/
-		tok = strtok_r(NULL, " ", &endp);
-		if (tok == NULL || *tok == '\0') {
-			return (wdns_res_parse_error);
+			/*
+			 * An unescaped comma separates two values. Add the
+			 * current 'v' to 'u' and reset it for the next value.
```
[the remark this answers is not in the document]
Comment (rafaelvanoni) [reply wdns/str_to_rdata_ubuf.c:329]: I thought we had converged on a reading of the spec where only alpn and keyNNN params could contain escape characters (?) I'm asking because I don't quite get the need for the first call to `rdata_from_str_string()` (the one you're suggesting, where we'd pass it the entire SvcParamValue string).

File: wdns/str_to_rdata_ubuf.c (hunk 8 above)
Comment (rafaelvanoni) [reply wdns/str_to_rdata_ubuf.c:280]: This chunk of the parse is changing, so no longer necessary.

File: wdns/str_to_rdata_ubuf.c (hunk 10 above)
Comment (rafaelvanoni) [reply wdns/str_to_rdata_ubuf.c:1085]: I use `strdup()` after identifying the `value` portion of `key=value`, before calling `str_to_svcparam()`. That said, about half the uses of `strpbrk()` in wdns also use `strdup()` and I'm not sure that saving on that small alloc/free is worth it. I also looked at implementing the `strtok_r()` pac-man style loop with `strpbrk()` and it wasn't as clean to me.

Comment (cmikk) [reply wdns/str_to_rdata_ubuf.c:1055]: Fair, and agreed that `strtok_r` is much more convenient when it works (which it does here, after `strdup()`).

Leaving this comment open until `dup` is no longer leaked on `str_to_svcparam` failure.

File: wdns/str_to_rdata_ubuf.c [hunk 13]
```diff
@@ -263,70 +233,116 @@ str_to_svcparam(ubuf *u, char *keyval)
...
-				return (wdns_res_parse_error);
-			}
+		dup = strdup(ubuf_cstr(v));
+		ubuf_destroy(&v);
 
-			ubuf_append(u, (uint8_t *)v6buf, sizeof (v6buf));
-			val_len += sizeof(v6buf);
-		} while ((tok = strtok_r(NULL, ",", &endp)) != NULL);
-		break;
+		len = strlen(dup);
```
Comment (cmikk) [inline wdns/str_to_rdata_ubuf.c:286]: Use `ubuf_data(v)` and `ubuf_size(v)` here instead of `strlen` on a `strdup`. If you want to destroy the ubuf `v` immediately afteward, `ubuf_detach` would also work.

The strdup/strlen approach would not parse an alpn value containing a zero byte (e.g. `alpn="b\000ar,baz"` would be parsed as `alpn="b"`), as it would be treated as a string terminator rather than part of the alpn value.

File: wdns/str_to_rdata_ubuf.c [hunk 14]
```diff
@@ -263,70 +233,116 @@ str_to_svcparam(ubuf *u, char *keyval)
...
+		for (i = 0; i < len; i++) {
+			char c = dup[i];
 
-		val_len = ubuf_size(u);
+			if (c == '\\') {
+				/* shouldn't happen, but just to be safe.. */
+				if (++i >= len) {
+					return (wdns_res_parse_error);
+				}
+				c = dup[i];
```
Comment (cmikk) [inline wdns/str_to_rdata_ubuf.c:304]: Low priority, but consider returning a parse error if character `i` is not `\` or `,` -- at this comma-separated list parsing layer, those are the only two permitted escaped characters.

File: t/test-rdata_to_str.c [hunk 15]
```diff
@@ -291,42 +291,129 @@ struct test tdata[] = {
...
+                    "\x00\x01"					/* alpn */
+                    "\x00\x0c"					/* length.. */
+                    "\x08"					/* ..value */
+                    "f\\oo,bar"
+                    "\x02"
+                    "h2",
+                .input_len = 35,
+                .rrclass = WDNS_CLASS_IN,
+                .rrtype = WDNS_TYPE_SVCB,
+                .expected = "16 foo.example.org. alpn=\"f\\\\oo\\,bar,h2\"",
```
Comment (cmikk) [inline t/test-rdata_to_str.c:381]: This needs to match:
```
		.input = "16 foo.example.org. alpn=\"f\\\\\\\\oo\\\\,bar,h2\"",
```
in the corresponding test case in `t/test-str_to_rdata.c`


In my estimation, we're nearing the point where C preprocessor hackery along the lines of:
```
#define QUOTE(...) #__VA_ARGS__
...
   .expected = QUOTE(16 foo.example.org. alpn="f\\\oo\\,bar,h2"),

```
is more ergonomic than manually stacking extra layers of escaping into C string literals. Note that this allows the argument to the `QUOTE` macro to exactly match the visible text from the standard, which may make this a net win.

(I've used the above hack to good effect in quoting JSON blobs as C literals. It works quite well as long as the quoted data does not include a `)` character)

File: wdns/str_to_rdata_ubuf.c [hunk 16]
```diff
@@ -263,70 +233,116 @@ str_to_svcparam(ubuf *u, char *keyval)
...
 
-	/*
-	 * Check that we processed all values (and that the ones who only
-	 * accept a single value don't have multiple).
-	 */
-	if (strtok_r(NULL, " ", &endp) != NULL) {
-		return (wdns_res_parse_error);
+		if (rdata_from_str_string((const uint8_t *)val, v) == 0) {
+			ubuf_destroy(&v);
+			fprintf(stderr, "ERRR (%s)\n", val);
```
Comment (cmikk) [inline wdns/str_to_rdata_ubuf.c:337]: This looks like a leftover debug print. Please remove.

File: t/test-rdata_to_str.c [hunk 17]
```diff
@@ -291,42 +291,129 @@ struct test tdata[] = {
...
+		    "\x02\x9b"					/* 667 */
+		    "\x00\x05"					/* length.. */
+		    "hello",					/* ..value */
+		.input_len = 28,
+		.rrclass = WDNS_CLASS_IN,
+		.rrtype = WDNS_TYPE_SVCB,
+		.expected = "1 foo.example.com. key667=\"hello\"",
+	},
+
+	/* appendix D, figure 5 is omitted as it fails the round trip test */
```
Comment (cmikk) [inline t/test-rdata_to_str.c:337]: This test can (and should) be restored, and (should) work.

**Review by cmikk — CHANGES_REQUESTED** [review]
The last round of changes have us very close to working and following the spec where needed. There are a few much smaller changes still needed, and then I think we will be ready for release.

File: wdns/str_to_rdata_ubuf.c (hunk 13 above)
Comment (rafaelvanoni) [reply wdns/str_to_rdata_ubuf.c:286]: Ah, good catch

File: wdns/str_to_rdata_ubuf.c (hunk 12 above)
[the remark this answers is not in the document]
Comment (cmikk) [reply wdns/str_to_rdata_ubuf.c]: Resolving this after an offline discussion summarized below:

The spec says that the params are treated as "character strings" in the zone file format. That said, all defined parameters other than alpn and keyNNN types have a natural and unambiguous representation which does not require quoting or escaping, so it's sufficient (for wdns) to skip `rdata_from_str_string`, as we don't need the latter's flexibility to faithfully represent any wire format value for those parameters.

**Review by cmikk — APPROVED** [review]
